File: crates/lib/superdev-core/assets/templates/web-react-android-ios-native/libs/native-debug-server-android/src/main/kotlin/com/nativedebugserver/DebugRouter.kt
// Route matching and dispatch for debug server endpoints.

package com.nativedebugserver

import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.time.format.DateTimeFormatter

/**
 * Parsed HTTP request.
 */
data class HttpRequest(
    val method: String,
    val path: String,
    val queryParams: Map<String, String>,
    val body: String?
) {
    val bodyJSON: JSONObject?
        get() = body?.takeIf { it.isNotBlank() }?.let {
            try { JSONObject(it) } catch (_: Exception) { null }
        }
}

/**
 * HTTP response to send.
 */
data class HttpResponse(
    val statusCode: Int,
    val contentType: String,
    val body: ByteArray
) {
    companion object {
        fun json(obj: JSONObject, status: Int = 200): HttpResponse {
            return HttpResponse(status, "application/json", obj.toString().toByteArray())
        }

        fun json(map: Map<String, Any?>, status: Int = 200): HttpResponse {
            return json(JSONObject(map), status)
        }

        fun png(data: ByteArray): HttpResponse {
            return HttpResponse(200, "image/png", data)
        }

        fun error(message: String, status: Int = 400): HttpResponse {
            return json(mapOf("error" to message), status)
        }

        val notFound = error("Not found", 404)
    }
}

/**
 * Routes incoming HTTP requests to appropriate handlers.
 */
class DebugRouter(private val logBuffer: LogBuffer) {
    private var appId: String = "unknown"

    private val lock = Any()
    private val stateProviders = mutableMapOf<String, () -> Map<String, Any?>>()
    private val actionHandlers = mutableMapOf<String, (Map<String, Any?>) -> Map<String, Any?>>()
    private var navigateHandler: ((String) -> Boolean)? = null
    private var screenshotProvider: (() -> ByteArray?)? = null

    fun configure(appId: String) {
        this.appId = appId
    }

    fun registerStateProvider(key: String, provider: () -> Map<String, Any?>) {
        synchronized(lock) { stateProviders[key] = provider }
    }

    fun registerAction(type: String, handler: (Map<String, Any?>) -> Map<String, Any?>) {
        synchronized(lock) { actionHandlers[type] = handler }
    }

    fun registerNavigateHandler(handler: (String) -> Boolean) {
        synchronized(lock) { navigateHandler = handler }
    }

    fun registerScreenshotProvider(provider: () -> ByteArray?) {
        synchronized(lock) { screenshotProvider = provider }
    }

    fun route(request: HttpRequest): HttpResponse {
        val parts = request.path.trimStart('/').split("/")
        if (parts.firstOrNull() != "debug") return HttpResponse.notFound

        val subPath = parts.drop(1)

        return when {
            request.method == "GET" && subPath == listOf("ping") -> handlePing()
            request.method == "GET" && subPath == listOf("logs") -> handleGetLogs(request.queryParams["since"])
            request.method == "POST" && subPath == listOf("logs", "clear") -> handleClearLogs()
            request.method == "GET" && subPath == listOf("screenshot") -> handleScreenshot()
            request.method == "GET" && subPath == listOf("state") -> handleGetAllState()
            request.method == "GET" && subPath.size == 2 && subPath[0] == "state" -> handleGetState(subPath[1])
            request.method == "GET" && subPath == listOf("routes") -> handleGetState("routes")
            request.method == "POST" && subPath == listOf("navigate") -> handleNavigate(request)
            request.method == "POST" && subPath == listOf("action") -> handleAction(request)
            else -> HttpResponse.notFound
        }
    }

    // MARK: - Core Handlers

    private fun handlePing(): HttpResponse {
        return HttpResponse.json(mapOf(
            "status" to "ok",
            "platform" to "android",
            "appId" to appId,
            "timestamp" to DateTimeFormatter.ISO_INSTANT.format(Instant.now())
        ))
    }

    private fun handleGetLogs(since: String?): HttpResponse {
        val sinceInstant = since?.let {
            try { Instant.parse(it) } catch (_: Exception) { null }
        }
        val entries = logBuffer.getEntries(sinceInstant)
        val entriesArray = JSONArray()
        entries.forEach { entriesArray.put(JSONObject(it.toMap())) }

        return HttpResponse.json(JSONObject().apply {
            put("entries", entriesArray)
            put("count", entries.size)
        })
    }

    private fun handleClearLogs(): HttpResponse {
        logBuffer.clear()
        return HttpResponse.json(mapOf("cleared" to true))
    }

    private fun handleScreenshot(): HttpResponse {
        val provider = synchronized(lock) { screenshotProvider }
        val data = provider?.invoke()
            ?: return HttpResponse.error("Screenshot capture not available", 500)
        return HttpResponse.png(data)
    }

    // MARK: - State Handlers

    private fun handleGetAllState(): HttpResponse {
        val providers = synchronized(lock) { stateProviders.toMap() }
        val state = mutableMapOf<String, Any?>()
        providers.forEach { (key, provider) -> state[key] = provider() }
        return HttpResponse.json(JSONObject(state as Map<String, Any?>))
    }

    private fun handleGetState(key: String): HttpResponse {
        val provider = synchronized(lock) { stateProviders[key] }
            ?: return HttpResponse.error("Unknown state key: $key", 404)
        return HttpResponse.json(JSONObject(provider()))
    }

    // MARK: - Action Handlers

    private fun handleNavigate(request: HttpRequest): HttpResponse {
        val body = request.bodyJSON ?: return HttpResponse.error("Missing 'route' in request body")
        val route = body.optString("route", "").takeIf { it.isNotEmpty() }
            ?: return HttpResponse.error("Missing 'route' in request body")

        val handler = synchronized(lock) { navigateHandler }
            ?: return HttpResponse.error("No navigate handler registered", 500)

        val success = handler(route)
        return HttpResponse.json(mapOf("navigated" to success, "route" to route))
    }

    private fun handleAction(request: HttpRequest): HttpResponse {
        val body = request.bodyJSON ?: return HttpResponse.error("Missing 'type' in request body")
        val type = body.optString("type", "").takeIf { it.isNotEmpty() }
            ?: return HttpResponse.error("Missing 'type' in request body")

        val payload = body.optJSONObject("payload")?.let { jsonToMap(it) } ?: emptyMap()

        val handler = synchronized(lock) { actionHandlers[type] }
            ?: return HttpResponse.error("Unknown action: $type", 404)

        val result = handler(payload)
        return HttpResponse.json(JSONObject(result as Map<String, Any?>))
    }

    private fun jsonToMap(json: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        json.keys().forEach { key -> map[key] = json.get(key) }
        return map
    }
}

File: crates/lib/superdev-core/assets/templates/web-react-android-ios-native/libs/native-debug-server-ios/Sources/NativeDebugServer/DebugRouter.swift
// Route matching and dispatch for debug server endpoints.

#if DEBUG
import Foundation

/// Parsed HTTP request.
struct HTTPRequest {
    let method: String
    let path: String
    let queryParams: [String: String]
    let body: Data?

    var bodyJSON: [String: Any]? {
        guard let body = body else { return nil }
        return try? JSONSerialization.jsonObject(with: body) as? [String: Any]
    }
}

/// HTTP response to send back.
struct HTTPResponse {
    let statusCode: Int
    let contentType: String
    let body: Data

    static func json(_ dict: [String: Any], status: Int = 200) -> HTTPResponse {
        let data = (try? JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys])) ?? Data()
        return HTTPResponse(statusCode: status, contentType: "application/json", body: data)
    }

    static func png(_ data: Data) -> HTTPResponse {
        HTTPResponse(statusCode: 200, contentType: "image/png", body: data)
    }

    static func error(_ message: String, status: Int = 400) -> HTTPResponse {
        json(["error": message], status: status)
    }

    static let notFound = error("Not found", status: 404)
}

/// Routes incoming HTTP requests to appropriate handlers.
final class DebugRouter: @unchecked Sendable {
    private let logBuffer: LogBuffer
    private var appId: String = "unknown"

    private let lock = NSLock()
    private var stateProviders: [String: @Sendable () -> [String: Any]] = [:]
    private var actionHandlers: [String: @Sendable ([String: Any]) -> [String: Any]] = [:]
    private var navigateHandler: (@Sendable (String) -> Bool)?
    private var screenshotProvider: (@Sendable () -> Data?)?

    init(logBuffer: LogBuffer) {
        self.logBuffer = logBuffer
    }

    func configure(appId: String) {
        self.appId = appId
    }

    func registerStateProvider(_ key: String, provider: @escaping @Sendable () -> [String: Any]) {
        lock.lock()
        stateProviders[key] = provider
        lock.unlock()
    }

    func registerAction(_ type: String, handler: @escaping @Sendable ([String: Any]) -> [String: Any]) {
        lock.lock()
        actionHandlers[type] = handler
        lock.unlock()
    }

    func registerNavigateHandler(_ handler: @escaping @Sendable (String) -> Bool) {
        lock.lock()
        navigateHandler = handler
        lock.unlock()
    }

    func registerScreenshotProvider(_ provider: @escaping @Sendable () -> Data?) {
        lock.lock()
        screenshotProvider = provider
        lock.unlock()
    }

    /// Route the request and return a response.
    func route(_ request: HTTPRequest) -> HTTPResponse {
        let pathComponents = request.path.split(separator: "/").map(String.init)

        // All routes are under /debug
        guard pathComponents.first == "debug" else {
            return .notFound
        }

        let subPath = Array(pathComponents.dropFirst())

        switch (request.method, subPath) {
        case ("GET", ["ping"]):
            return handlePing()

        case ("GET", ["logs"]):
            return handleGetLogs(since: request.queryParams["since"])

        case ("POST", ["logs", "clear"]):
            return handleClearLogs()

        case ("GET", ["screenshot"]):
            return handleScreenshot()

        case ("GET", ["state"]):
            return handleGetAllState()

        case ("GET", _) where subPath.count == 2 && subPath[0] == "state":
            return handleGetState(key: subPath[1])

        case ("GET", ["routes"]):
            return handleGetState(key: "routes")

        case ("POST", ["navigate"]):
            return handleNavigate(request)

        case ("POST", ["action"]):
            return handleAction(request)

        default:
            return .notFound
        }
    }

    // MARK: - Core Handlers

    private func handlePing() -> HTTPResponse {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return .json([
            "status": "ok",
            "platform": "ios",
            "appId": appId,
            "timestamp": formatter.string(from: Date())
        ])
    }

    private func handleGetLogs(since: String?) -> HTTPResponse {
        var sinceDate: Date? = nil
        if let since = since {
            let formatter = ISO8601DateFormatter()
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            sinceDate = formatter.date(from: since)
        }
        let entries = logBuffer.getEntries(since: sinceDate)
        return .json([
            "entries": entries.map { $0.toDict() },
            "count": entries.count
        ])
    }

    private func handleClearLogs() -> HTTPResponse {
        logBuffer.clear()
        return .json(["cleared": true])
    }

    private func handleScreenshot() -> HTTPResponse {
        lock.lock()
        let provider = screenshotProvider
        lock.unlock()

        guard let provider = provider, let data = provider() else {
            return .error("Screenshot capture not available", status: 500)
        }
        return .png(data)
    }

    // MARK: - State Handlers

    private func handleGetAllState() -> HTTPResponse {
        lock.lock()
        let providers = stateProviders
        lock.unlock()

        var state: [String: Any] = [:]
        for (key, provider) in providers {
            state[key] = provider()
        }
        return .json(state)
    }

    private func handleGetState(key: String) -> HTTPResponse {
        lock.lock()
        let provider = stateProviders[key]
        lock.unlock()

        guard let provider = provider else {
            return .error("Unknown state key: \(key)", status: 404)
        }
        return .json(provider())
    }

    // MARK: - Action Handlers

    private func handleNavigate(_ request: HTTPRequest) -> HTTPResponse {
        guard let body = request.bodyJSON, let route = body["route"] as? String else {
            return .error("Missing 'route' in request body")
        }

        lock.lock()
        let handler = navigateHandler
        lock.unlock()

        guard let handler = handler else {
            return .error("No navigate handler registered", status: 500)
        }

        let success = handler(route)
        return .json(["navigated": success, "route": route])
    }

    private func handleAction(_ request: HTTPRequest) -> HTTPResponse {
        guard let body = request.bodyJSON, let type = body["type"] as? String else {
            return .error("Missing 'type' in request body")
        }

        let payload = body["payload"] as? [String: Any] ?? [:]

        lock.lock()
        let handler = actionHandlers[type]
        lock.unlock()

        guard let handler = handler else {
            return .error("Unknown action: \(type)", status: 404)
        }

        let result = handler(payload)
        return .json(result)
    }
}

#endif
